src/Pos/Block/Slog/Logic.hs

```diff
@@ -303,8 +307,16 @@ slogRollbackBlocks blunds = do
         mconcat [forwardLinksBatch, inMainBatch]
 
 -- Common actions for rollback and apply.
-slogCommon :: MonadSlogApply ssc ctx m => LastBlkSlots -> m ()
+slogCommon
+    :: MonadSlogApply ssc ctx m
+    => LastBlkSlots
+    -> m ()
 slogCommon newLastSlots = do
     sanityCheckDB
     slogPutLastSlots newLastSlots
-    putSlottingData =<< GS.getSlottingData
+    -- We read from the database and write in the memory.
+    -- TODO(ks): This is unsafe! We don't have control over sequentiality and
+    -- use explicit indexing. It would be better if we have sorted EpochSlotData and
+    -- pass it without the index.
```

Comment:
I don't know. It looks messed up to me. For example, `slogCommon` is called both from _apply_ and _rollback_. 
So in effect no matter what happens, either an _apply_ or _rollback_, it always reads `SlottingData` from database and writes into memory? That seems wrong. What effect will that have when a _rollback_ happens?
If what you said about apply holds (only a single epoch can be computed at a time) and on rollback doesn't, it gets quite complicated.
That's why I don't want to change it more than I have to.